verifier/newverifier.go

```diff
@@ -0,0 +1,45 @@
+package verifier
+
+import (
+	"bytes"
+	"fmt"
+	"github.com/spacemeshos/merkle-tree"
+	"github.com/spacemeshos/poet/shared"
+	"sort"
+)
+
+func Validate(proof shared.MerkleProof, challenge shared.NewChallenge, leafCount uint64, securityParam uint8) error {
+	if int(securityParam) != len(proof.ProvenLeaves) {
+		return fmt.Errorf("number of proven leaves (%d) must be equal to security param (%d)",
+			len(proof.ProvenLeaves), securityParam)
+	}
+	provenLeafIndices := asSortedSlice(shared.FiatShamir(proof.Root, leafCount, securityParam))
+	valid, parkingSnapshots, err := merkle.ValidatePartialTreeWithParkingSnapshots(provenLeafIndices,
+		proof.ProvenLeaves, proof.ProofNodes, proof.Root, challenge.MerkleHashFunc())
+	if err != nil {
+		return fmt.Errorf("error while validating merkle proof: %v", err)
+	}
+	if !valid {
+		return fmt.Errorf("merkle proof not valid")
+	}
+
+	for id, label := range proof.ProvenLeaves {
+		expectedLabel := shared.MakeLabel(challenge.LabelHashFunc(), provenLeafIndices[id], parkingSnapshots[id])
+		if !bytes.Equal(expectedLabel, label) {
+			return fmt.Errorf("label at index %d incorrect", id)
+		}
+	}
+
+	return nil
+}
+
+func asSortedSlice(s map[uint64]bool) []uint64 {
+	var ret []uint64
+	for key, value := range s {
```

Comment:
Why do you pass it as a dictionary to begin with? Isn't it better if the FiatShamir method will return a set instead of a map?

Reply:
Note that `merkle.GenerateProof()` expects it that way (`map[uint64]bool`). I assume that by "set" you mean a sorted slice?